torch/csrc/jit/codegen/cuda/scheduler/registry.cpp

```diff
@@ -358,6 +358,40 @@ class SchedulerTopologyChecker {
 
     return true;
   }
+
+  // Returns if any non-trivial views are a dependency of the expression
+  // producing the reference. This could be important as transform propagation
+  // from a reference backwards through a view should always work, but transform
+  // propagation form a reference forward through a view could interfere with
+  // the view transforms.
+  static bool hasViewNotDependentOnRef(
```

Comment:
Yes we're checking statement 2, you're right it should be flipped. I think it has equivalent safety. The only time I think this would have better safety is in the case you mentioned, but in that case I think we'd already fail at some point because we have dimensions in T1 that map to dimensions in T1.
 
For multiple view cases I think this approach should be fine and I don't think the propagation approach provides much, especially if we have multiple views. If you pass through one view, how do you know when you come back up through two views it's a consistent format.

The way I'm trying to think about this is views are format transformations, and operations that connect through non-view's are the "same format". Format in this sense is size and number of the dimensions.

The only way I can think to "lose" information and propagate in a bad way through a view with this check is if we propagate through reductions from output->input direction. However our reduction schedulers all use a reduction as a reference, so that shouldn't be an issue.

I think dependency is the right check, and I could write a new dependency check in less complexity. Again the only issue I can see here is if we have the pattern you mention above because I don't know how else you can get around a view without another consistent view. However, I think we should avoid these self-ID-mapping cases as mentioned in https://github.com/csarofeen/pytorch/pull/1941 so we should really make sure we catch those instances in segmentation.